tsl/src/chunk_api.c

```diff
@@ -0,0 +1,1498 @@
+							CSTRING_TYPEALIGN);
+	values[AttrNumberGetAttrOffset(Anum_chunk_colstats_slot_valtype_strings)] =
+		PointerGetDatum(array);
+}
+
+/*
+ * Construct a tuple for the get_chunk_relstats SQL function.
+ */
+static HeapTuple
+chunk_get_single_colstats_tuple(Chunk *chunk, int column, TupleDesc tupdesc)
+{
+	HeapTuple ctup;
+	Form_pg_statistic pgsform;
+	Datum values[_Anum_chunk_colstats_max];
+	bool nulls[_Anum_chunk_colstats_max] = { false };
+	bool dropped;
+
+	if (DatumGetBool(DirectFunctionCall1(row_security_active, ObjectIdGetDatum(chunk->table_id))))
+		return NULL;
+
+	ctup = SearchSysCache2(ATTNUM, ObjectIdGetDatum(chunk->table_id), column);
+	if (!HeapTupleIsValid(ctup))
+		return NULL;
+
+	dropped = ((Form_pg_attribute) GETSTRUCT(ctup))->attisdropped;
+	ReleaseSysCache(ctup);
+
+	if (dropped)
+		return NULL;
+
+	if (!DatumGetBool(DirectFunctionCall3(has_column_privilege_id_attnum,
+										  ObjectIdGetDatum(chunk->table_id),
+										  Int16GetDatum(column),
+										  PointerGetDatum(cstring_to_text("SELECT")))))
+		return NULL;
+
+	ctup = SearchSysCache3(STATRELATTINH, ObjectIdGetDatum(chunk->table_id), column, false);
+
+	/* pg_statistics will not have an entry for an unanalyzed table */
+	if (!HeapTupleIsValid(ctup))
+		return NULL;
+
+	pgsform = (Form_pg_statistic) GETSTRUCT(ctup);
+
+	values[AttrNumberGetAttrOffset(Anum_chunk_colstats_chunk_id)] = Int32GetDatum(chunk->fd.id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_colstats_hypertable_id)] =
+		Int32GetDatum(chunk->fd.hypertable_id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_colstats_column_id)] = Int32GetDatum(column);
+	values[AttrNumberGetAttrOffset(Anum_chunk_colstats_nullfrac)] =
+		Float4GetDatum(pgsform->stanullfrac);
+	values[AttrNumberGetAttrOffset(Anum_chunk_colstats_width)] = Int32GetDatum(pgsform->stawidth);
+	values[AttrNumberGetAttrOffset(Anum_chunk_colstats_distinct)] =
+		Float4GetDatum(pgsform->stadistinct);
+
+	collect_colstat_slots(ctup, pgsform, values, nulls);
+
+	ReleaseSysCache(ctup);
+
+	return heap_form_tuple(tupdesc, values, nulls);
+}
+
+static void
+chunk_update_colstats(Chunk *chunk, int16 attnum, float nullfract, int32 width, float distinct,
+					  ArrayType *kind_array, Oid *slot_ops, ArrayType **slot_numbers,
+					  Oid *value_kinds, ArrayType **slot_values)
+{
+	Relation rel;
+	Relation sd;
+	Datum values[Natts_pg_statistic];
+	bool nulls[Natts_pg_statistic];
+	bool replaces[Natts_pg_statistic];
+	HeapTuple stup;
+	HeapTuple oldtup;
+	int i, k;
+	int *slot_kinds;
+
+	rel = try_relation_open(chunk->table_id, ShareUpdateExclusiveLock);
+
+	/* If a vacuum is running we might not be able to grab the lock, so just
+	 * raise an error and let the user try again. */
+	if (NULL == rel)
+		ereport(ERROR,
+				(errcode(ERRCODE_LOCK_NOT_AVAILABLE),
+				 errmsg("unable to aquire table lock to update column statistics on \"%s\"",
+						NameStr(chunk->fd.table_name))));
+
+	sd = relation_open(StatisticRelationId, RowExclusiveLock);
```

Comment:
This code was partly taken from PG and I think the reason this is using `relation_open` is because there's a `try_relation_open` above, and there is no `try_table_open` AFAIK. In any case, `relation_open` is a valid way of opening the relation (which is actually what `table_open` ultimately does, AFAIK.